File: src/net/dinkla/raytracer/math/Vector2D.scala
package net.dinkla.raytracer.math

/**
 * Created by Dinkla on 10.07.2015.
 */
abstract class Vector2D[@specialized(Float, Double) F](x: F, y: F)(implicit numType: Fractional[F]) extends Element2D[F](x, y) {

  def unary_-(): Vector2D[F]           // unary minus

  def +(v: Vector2D[F]): Vector2D[F]        = plus(v)     // Vector addition gives Vector
  def plus(v: Vector2D[F]): Vector2D[F]     // Vector addition gives Vector

  def -(v: Vector2D[F]): Vector2D[F]        = plus(-v)  // Vector addition gives Vector
  def minus(v: Vector2D[F]): Vector2D[F]    = plus(-v)  // Vector addition gives Vector

  def dot(v: Vector2D[F]): F                // dot product

  def toPoint: Point2D[F]

  def normalize: Vector2D[F]                // the normal vector with length 1

  override def equals(obj: scala.Any): Boolean
  = obj match {
    case that: Vector2D[F] => x == that.x && y == that.y
    case _ => false
  }

  override def hashCode(): Int = (x, y).hashCode()

}

object Vector2D {

  def Vector2D(x: Double, y: Double) = new Vector2DD(x, y)
  def Vector2D(x: Float, y: Float) = new Vector2DF(x, y)

}

File: src/main/kotlin/net/dinkla/raytracer/math/Vector2D.kt
package net.dinkla.raytracer.math

class Vector2D(x: Double, y: Double) : Element2D(x, y) {

    operator fun plus(v: Vector2D) = Vector2D(x + v.x, y + v.y)

    operator fun minus(v: Vector2D) = Vector2D(x - v.x, y - v.y)

    operator fun times(s: Double) = Vector2D(s * x, s * y)

    infix fun dot(v: Vector2D): Double = x * v.x + y * v.y

    infix fun dot(v: Normal): Double = x * v.x + y * v.y

    fun normalize(): Vector2D {
        val l = length()
        return Vector2D(x / l, y / l)
    }

}

operator fun Double.times(v: Vector2D) = Vector2D(this * v.x, this * v.y)
